### ISSUE TITLE
[css-overflow-3] Clarify padding in overflow content

### ISSUE THREAD

**Issue by jcrben** [w3c/csswg-drafts#129]
I discovered this today when my overflow content appeared cut-off in Firefox and IE. It appears this was reported to Mozilla and closed as invalid in [Bug 748518 - padding-bottom/right(block-end/inline-end) is ignored with overflow:auto/scroll because it extends in from the border-box rather than out from the scrollable area](https://bugzilla.mozilla.org/show_bug.cgi?id=748518) and someone there suggested bringing it up with the CSS team. I did a [search for padding-bottom overflow on the www-style list](https://www.w3.org/Search/Mail/Public/search?keywords=padding-bottom+overflow&hdr-1-name=subject&hdr-1-query=&index-grp=Public_FULL&index-type=t&type-index=www-style) and didn't see anything.

Previous practical discussions on Stack Overflow include [Bottom padding not working on overflow element in non-chrome browsers](http://stackoverflow.com/questions/10722367/bottom-padding-not-working-on-overflow-element-in-non-chrome-browsers) and [padding-bottom is ignored in Firefox and IE on overflowing elements with no content](http://stackoverflow.com/questions/20624938/padding-bottom-is-ignored-in-firefox-and-ie-on-overflowing-elements-with-no-cont).

Perhaps this is already clear and Chrome has it wrong? I'll admit I'm not enough of an expert on the spec to be sure, just hoping for consistency.

**Comment by tabatkins** [comment]
It was never clear in 2.1.  Overflow 3 is clear about it - [padding should be included](https://drafts.csswg.org/css-overflow-3/#overflow-concepts) - but it's not 100% clear that's web-compatible.  In particular, Chrome only includes padding in the *block* axis, and as you found, Firefox/IE don't include it in *either* axis. We need to dig thru this as a group and figure out what we *should* be doing.

**Comment by tabatkins** [comment]
We discussed this on the call today. I've added an issue to the spec for it.

**Comment by litherum** [comment]
In WebKit, we actually have different policies for block children and inline children. If the scroll container has inline children, we will include padding-right (or the equivalent for different writing modes) when computing layout overflow. However, if the scroll container has block children, we just consider the border boxes of the children. This is why there was some disagreement in the call today about what the policy is for the inline-direction.

**Comment by frivoal** [comment]
Since additional information is being added here, and we haven't solved the original issue, I'm reopening this despite also having an inline issue in the spec.

**Comment by frivoal** [comment]
@litherum Very interesting. Do you know if the decision to behave differently in the inline direction was informed by compat issues? I would not be surprised that adding the padding in the inline direction if the scroll container has block children was causing scroll bars to appear in places where they're not wanted, and had to be avoided because of that.

**Comment by tabatkins** [comment]
@litherum I thought it might be something like that; I'm glad to hear it's that simple, and I'll edit the issue accordingly.  I echo @frivoal's question about whether the "don't include if if there's only block children" was done for compat (y'all were seeing pages suddenly sprout lots more scrollbars than intended), or if it was just happenstance (y'all added a hack to make inline content look better, and just didn't apply it generally).

**Comment by tabatkins** [comment]
@litherum Ping on:

> I echo @frivoal's question about whether the "don't include if if there's only block children" was done for compat (y'all were seeing pages suddenly sprout lots more scrollbars than intended), or if it was just happenstance (y'all added a hack to make inline content look better, and just didn't apply it generally).

**Comment by litherum** [comment]
This code is older than my time on the WebKit team. I'll ask tomorrow.

**Comment by sarazhang123** [comment]
@tabatkins Do you know whether firefox will fix this bug https://bugzilla.mozilla.org/show_bug.cgi?id=748518 ?

**Comment by SebastianZ** [comment]
> @tabatkins Do you know whether firefox will fix this bug https://bugzilla.mozilla.org/show_bug.cgi?id=748518 ?

1. The [spec. issue](https://drafts.csswg.org/css-overflow-3/#issue-57abf768) is not resolved yet. And I assume Mozilla won't reopen that bug report until the spec. issue got resolved.
2. Tab is working for Google, so he's probably the wrong person to ask about Firefox implementation. 😉 But @dholbert, @dbaron, or @bzbarsky might be able to provide feedback here.

Sebastian

**Comment by bzbarsky** [comment]
My two cents are that we will change behavior here no more than once.  That means the WG needs to get its act together and figure out what it's doing before we make any changes.

**Comment by nuxodin** [comment]
From a web developers view the Chrome behavior makes much more sense.
I do not like it to put semantic-less wrapper-divs to get my padding.

**Comment by frivoal** [comment]
@litherum On June 22, 2016, you said:
>  This code is older than my time on the WebKit team. I'll ask tomorrow.

Unless you live in a very unusual time zone, I think we're past tomorrow :) Any update?

**Comment by frlinw** [comment]
1) Hello, I'm a developer and I lost a precious time to understand why there is padding-top/left/right but no bottom on firefox/edge.
2) The spec should be clear about this, display all paddings or hide all paddings. The "show padding-bottom/right if...." leads to 1)

Gecko is close to the spec (and I like it, most of the time) but the webkit implementation makes more sense in this case.

**Comment by fantasai** [comment]
@frlinw I think historically the issue is that browsers didn't want to trigger scrollbars for `overflow: auto` unless visible was overflowing the inner border edge, so they didn't count padding. Once scrollbars are in place, however, it makes sense that the padding is included in the scrollable area, otherwise (as we've all noticed) the content gets uncomfortably tight against the bottom right edge when we scroll to that corner.

Unfortunately, browsers only measure overflow in one way: either include a padding/margin or don't. And changing behavior to include padding/margin when it wasn't before can cause pages which previously didn't have scrollbars to suddenly have scrollbars, and this can make a problem for the page. So we're in an uncomfortable position of what to do here.

I think it would be best artistically if we can count padding/margin as much as possible, but compatibility with existing Web content limits what we can do about it...

**Comment by frlinw** [comment]
> can cause pages which previously didn't have scrollbars to suddenly have scrollbars

Yes, I understand the issue here but actually, it is expected. Padding can cause overflow.
What is unexpected is browsers trying to workaround this expected behavior with an unexpected one.
If you go this way, padding-top can cause overflow when it's unecessary why the spec didn't decide to ignore padding-top too ? Because it breaks something, and it will be very visible. It's the same thing with padding-bottom, if you ignore it, even if it's less visible, it breaks something.

Moreover, Blink has the highest market share, it does not implement the behavior introduced in CSS 2.1 and there is no screams about it (actually maybe there are screams, I really don't know, but blink team does not need 6 years to implement something the dev community wants).

The issue here is the developer is not in control. The spec introduced a behavior which can be unexpected but didn't give the developer the opportunity to control it (with something like `overflow-clip`?)

>  existing Web content limits what we can do

CSS 2.1 broke existing web content because of this change too. Web is a living thing.

**Comment by nuxodin** [comment]
To solve the Web-Compat problem one could invent a CSS property such:
`overflow-padding-behavior-right: strict;`
`overflow-padding-behavior-bottom: strict;`
And a shorthand (right bottom):
`overflow-padding-behavior: strict strict;`

values are:
- `strict` => padding at the end
- `inline-content` or inline-children ? => padding-only for inline-children
- `ignore` => no padding at the end

The actual Chrome behavior would be:
`overflow-padding-behavior: inline-content strict;`

The actual Firefox behavoir would be:
`overflow-padding-behavior: ignore ignore;`

**Comment by frlinw** [comment]
@nuxodin your proposition seems good. I'm in for a css property to give some control to web authors (if there is no consensus about a default behavior) but it is not really related to the webcompat issue
Because your spec is supposed to give a default value for this property... So implement this spec will break (or repair, depends of the point of view) some current browsers behavior

**Comment by fantasai** [comment]
So, due to some issues in the css-align-3 spec (see https://github.com/w3c/csswg-drafts/issues/1425), the `align-content` and `justify-content` properties will now have this side-effect: any non-`normal` value will cause the padding to be honored as requested here. See https://www.w3.org/TR/css-align-3/#overflow-scroll-position I think that resolves the request for a switch. :) We should definitely note this in the Overflow spec as well, however; and this issue remains open to see what we can do to fix this in the default case (which is constrained by Web-compat). I suspect we might be able to make it work in the block axis, but not the inline one.

**Comment by ashishjat** [comment]
@jcrben - still no solution found for the above problem , but we can try a trick to work. create a empty div at the bottom of your scrolling div and give it some height.this will work very similar to  padding bottom.

**Comment by fantasai** [comment]
OK, so we're *also* going to try to fix this for Flexbox and Grid generally (even when alignment is normal). See https://github.com/w3c/csswg-drafts/issues/3665 Cross your fingers and hope it's Web-compatible. :)

**Comment by OliverJAsh** [comment]
For anyone who comes here looking for a workaround/solution, here is a very well documented workaround:

https://www.brunildo.org/test/overscrollback.html

Problem: https://jsbin.com/sawuwor/1/edit?html,css,output
Workaround: https://jsbin.com/wikuyif/1/edit?html,css,output

**Comment by jipema** [comment]
Another workaround that worked well in my case: adding bottom margin to the last child.
```
 .overflowing-container > *:last-child{ margin-bottom: 20px; }
```

**Comment by silverwind** [comment]
Something should be done here and I think it should be to change the spec to what Blink implements given that most developers would expect Blink/Webkit behaviour as the "correct" one.

https://bugzilla.mozilla.org/show_bug.cgi?id=748518 has accumulated 21 dupe bugs since it was opened 8 years ago, which shows developers have come to expect Blink/Webkit as the correct one and Firefox's as wrong.

It certainly seems intuitive to me that padding/margin should not dissappear just because a area is scrollable. I would visualize the scroll area as a "zoom-in" on a content area. A "zoom" should not be able to change content.

**Comment by lucijanblagonic** [comment]
My workaround for fixing this bug in my project was (helpful when you want the children centered with `margin: auto` in a flex container if there is no scroll.):

.overflowing-container > *:last-child:after {
    display: block;
    height: 1rem; /* Match bottom padding */
    margin-bottom: -1rem; /* Move it outside */
    content: "";
}

**Comment by rahulthewall** [comment]
Ran into this issue today, thankfully I was able to see this discussion. Added a margin-bottom to the div inside the scroll div and fixed the issue.

**Comment by radogado** [comment]
After a few days of fighting this problem, today I filed a bug https://bugzilla.mozilla.org/show_bug.cgi?id=1621708 and realised it's an 8yo known issue. Used the margin workaround, but now I can't have a flex container. Shame.

**Comment by lkraav** [comment]
> After a few days of fighting this problem, today I filed a bug https://bugzilla.mozilla.org/show_bug.cgi?id=1621708 and realised it's an 8yo known issue. Used the margin workaround, but now I can't have a flex container. Shame.

Yep, I also filed yet another duplicate for it just a few days earlier.

Fortunately managed to find a solid x-browser workaround with making the scroll container a child of the original scroll container. With custom elements, `<slot>` element was already sitting there, I was able to to utilize it directly for become the new scroll container, instead of creating another `<div>` layer or similar. Implementation is live at https://cxl.com/blog/ right sidebar.

**Comment by radogado** [comment]
How can fundamental properties like padding and scroll be broken (sorry, "undefined") since forever, while the W3C is pushing "magazine layout with CSS Grid" etc.? Please, let's fix the fundamentals and have a solid CSS before rushing to the next hot thing. Thanks.

**Comment by radogado** [comment]
Workarounds with pseudo elements don't work when I need to set the padding to percentage. Please fix it or add an option to support end padding. Thanks.

**Comment by bfgeek** [comment]
I'm about to add use-counters into blink to determine how often we have overflow-x + inline-end padding. We should have data in 2-3 months.

Ian

**Comment by bfgeek** [comment]
Over the past few weeks I've done a fair bit of investigation on the model here. Broadly speaking there is the "ideal" model which is as follows.

When calculating the scrollable-overflow for a scroll-container there are two rectangles we need to keep track of:

[1] Inflow bounds
--------------

To determine this correctly the engine needs to keep track of the bounds of all *inflow* children and their margins (pre relative-positioning and transform adjustments).
Padding of the scroll container is then added to this rectangle and added to the scrollable-overflow.

Importantly this *doesn't* include any float, or out-of-flow positioned children.

This also only happens for a scroll container, this *isn't* part of the scrollable-overflow for non-scroll containers.

[2] Scrollable-overflow bounds
---------------------------

On top of this the "regular" scrollable overflow calculations apply, e.g. the scrollable-overflow for propagation rectangle which is post-transform, post-relative-positioning.

The result of the scrollable overflow rectangle is the union of [1], [2].

Web-compat concerns
-----------------------

Broadly speaking everything is a relative mess :). However the compat concerns only surround [1].

I created this handle table!
|                           | Blink (inline) | Blink (block)          | WebKit (inline) | WebKit (block)         | Gecko (inline) | Gecko (block) | EdgeHTML (inline) | EdgeHTML (block) |
|---------------------------|----------------|------------------------|-----------------|------------------------|----------------|---------------|-------------------|------------------|
| [block-flow (inline-level)](https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=8499) | padding        | padding                | padding         | padding                | none           | none          | none              | none             |
| [block-flow (block-level)](https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=8496)  | none           | child-margin + padding | none            | child-margin + padding | none           | child-margin  | none              | child-margin     |
| [flex](https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=8497)                      | none           | child-margin + padding | none            | child-margin + padding | none           | none          | none              | none             |
(I didn't test grid).
The "ideal" thing here (which developers are asking for) would be for each of the layout modes to include "child-margin + padding".

(For the "block-flow (inline-level)" case "child-margin + padding" == "padding").

What to do
-----------

We (Blink) are prepared to try and incrementally move to the "child-margin + padding" case for both axis, however stop when we find real world compat concerns. As an example this may end up as the following:
- flex includes child-margin+padding.
- block-flow (block-level) stays as is.
(or visa versa).

We've got some UseCounters in place, but we are in the process of adding more precise ones.

**Comment by bfgeek** [comment]
@fantasai The primary difference from the spec and above, is that the padding-edges contributing to scrollable-overflow are determined "pre-transform, pre-relative-position". E.g. this means that you can animate an element "in" from the top for example, without affecting the scrollable-overflow. This is obviously still union'ed if a child is transformed "outside" this area.

This is what Blink/WebKit do today for block-end edges (and what grid somewhat requires, making the grid part of the scrollable-area).

**Comment by bfgeek** [comment]
Additionally OOF-positioned children aren't considered for the padding calculation.

**Comment by bfgeek** [comment]
@fantasai & I chatted quicky and are on the same page regarding my previous two comments.

**Comment by davidhmays** [comment]
> Another workaround that worked well in my case: adding bottom margin to the last child.
> 
> ```
>  .overflowing-container > *:last-child{ margin-bottom: 20px; }
> ```

This would have been a great workaround for me as well, until I had dynamically created cards in a grid. This left the last child having a nice look, but if there were several children next to each other, only having the last one have padding threw it all off.

**Comment by aethanyc** [comment]
I'm [fixing this webcompat issue](https://groups.google.com/g/mozilla.dev.platform/c/J9O9oT6Y64U/m/P1yfpBysAQAJ) for gecko. I update @bfgeek's compat table in the [previous comment](https://github.com/w3c/csswg-drafts/issues/129#issuecomment-697691447) based my test. Please let me know if there's any error.
| Testcase                                                                                                    | Blink (inline-end)     | Blink (block-end)      | WebKit (inline-end) | WebKit (block-end)     | Gecko (inline-end)     | Gecko (block-end)      |
|-----------------------------------------------------------------------------------------------------|------------------------|------------------------|---------------------|------------------------|------------------------|------------------------|
| [block-flow (inline-level)](https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=8499) | padding                | padding                | padding             | padding                | none [3]               | padding                |
| [block-flow (block-level)](https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=8496)  | none                   | child-margin + padding | none                | child-margin + padding | none [3]               | child-margin + padding |
| [flex](https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=8497)                      | child-margin + padding | child-margin + padding | none                | child-margin + padding | child-margin + padding | child-margin + padding |
| [grid](https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=9117)                      | none [1]               | none [1]               | none                | child-margin + padding | none [2]               | none [2]               |

(For the "block-flow (inline-level)" case "child-margin + padding" == "padding").

* Blink (Google Chrome 91.0.4459.0)
* WebKit (Safari Technology Preview, Safari 14.2, WebKit 16612.1.6.2)
* Gecko (Firefox Nightly 89, 2021-03-26)
* Omit EdgeHTML because I don't test it.

[1] Blink is planning to change grid to "child-margin + padding" in both axes in GridNG (from my communication with @bfgeek)
[2] Gecko is planning to change grid to "child-margin + padding" in both axes, too. Tracking in [Bug 1527539](https://bugzilla.mozilla.org/show_bug.cgi?id=1527539).
[3] Both Blink and WebKit are not consistent on whether the inline-end padding is added in block-flow (yes for inline-level, no for block-level), so Gecko is waiting until there's resolution for this issue. Tracking in [Bug 1700858](https://bugzilla.mozilla.org/show_bug.cgi?id=1700858).

**Comment by fantasai** [comment]
> The primary difference from the spec and above, is that the padding-edges contributing to scrollable-overflow are determined "pre-transform, pre-relative-position".

@bfgeek If the spec isn't clear about that, then that's a problem with the spec that should be fixed. It's certainly intending to specify that. Padding gets applied essentially at the edge of what would have been the content edge of the box, if it were auto-sized to its contents. Are you saying the spec is unclear, or that implementations attach padding after transforms/relpos rather than before?

> Additionally OOF-positioned children aren't considered for the padding calculation.

"Out of flow" includes floats. I think floats are intended to be included here, no?

**Comment by franktopel** [comment]
And in the meantime [another question about this has been raised and suggested quirky workarounds on StackOverflow](https://stackoverflow.com/questions/66984685/how-can-i-ensure-padding-works-on-a-grid-item-with-limited-height).

**Comment by radogado** [comment]
> Considering [this issue has been reported on bugzilla more than 9 (!!!!) years ago](https://bugzilla.mozilla.org/show_bug.cgi?id=748518), to an outside person this does make the impression that whoever decides on such issues doesn't do their job in anywhere near acceptable timely manner.

Must be said again. Frontenders are disheartened by browser bugs, quirks and undefined behaviour. From all-time favourites like scroll padding and missing subpixel scroll offset to relatively new unstable stuff like scroll snapping, web dev is a minefield.

**Comment by frivoal** [comment]
As far as I can tell, other than possible editorial tweaks to make the prose cleaner, the only reminaing open quesiton is that of “[3]” in https://github.com/w3c/csswg-drafts/issues/129: do we add that padding in the inline-end of block-flow or not (limited to the case of justify-content: normal).

I think that:
* adding it is nicer in terms of readibility / spacing, as it allows content not to smash in the border even though you specified padding.
* adding it in places where it wasn't included before is likely to cause scrollbars to appear where they were none before, and that would be considered unacceptable breakage

That would mean that even though the current webkit/blink behavior is odd, it represents the most padding we'll ever get, so we might want to spec that.

However, the blink/webkit "odd" behavior is *really* odd: they  add the scroll container’s padding to the inline-end of inline *children* of the container, but not to other descendents of the container, be they block containers or inline descendents other than children.

This is not just weird, it also violates the principle that inserting an unstyled div should make no difference to the layout.

We probably should for consistency, and never add that padding to the inline-end.

**Comment by aethanyc** [comment]
http://wpt.live/css/css-overflow/overflow-padding.html should be fixed to match the resolution.

**Comment by frivoal** [comment]
@bfgeek Can you say a little bit about why you want this back on the agenda? Should I hold off doing the edits?

**Comment by bfgeek** [comment]
Yes - sorry I was OOO last week. I'm pretty sure that we can do something better for web developers here but we'd need to un-resolve that resolution. We aren't necessarily compat constrained here. (have data).

**Comment by keviinnch** [comment]
umm, I'm sorry, but is this issue resolved? I use Chrome 92.0.4515.131 and it seems that the padding (right-padding) is already work as expected, or this thread is only for discussing the draft spec, but not the issue itself? (sorry if this is a dumb question..)

**Comment by frivoal** [comment]
>  this thread is only for discussing the draft spec

Yes, this whole repository is for discussing specifications. Of course, specs and implementations are expected to match each-other, but if you're interested in the status of a particular bug in a particular browser, this is the wrong repo to be looking at.

**Comment by lyjk-underdog** [comment]
If the scroll container has different types of children, how to determine the scrollable overflow rectangular area?

**Comment by dholbert** [comment]
@bfgeek, just wanted to check whether there are any updates / timeline-ETAs on the proposed use-counter/webcompat-data-gathering phase, for determining whether your proposal is web-compatible? (per https://github.com/w3c/csswg-drafts/issues/129#issuecomment-893052191 )

This is an area of occasional interop pain (just got a bug report in https://bugzilla.mozilla.org/show_bug.cgi?id=1744414 ) so I'm looking forward to having a resolution for something web-compatible and interoperable. :)

**Comment by bfgeek** [comment]
@dholbert - Getting there. We successfully shipped adding inline-end padding to blocks which *already* scrolled. E.g.

```
<!DOCTYPE html>
<div style="overflow: scroll; padding: 10px; border: solid; width: 100px; height: 100px;">
  <div style="width: 100px; height: 100px; background: lime;"></div>
</div>
```
https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=9874

We are still determining if we can add this inline-end padding universally. There is one interesting subcase which is highly prevalent:
```
<!DOCTYPE html>
<div style="overflow: scroll; padding: 10px; border: solid; width: 100px; height: 100px;">
  <div style="width: 100px; height: 0; background: lime;"></div>
</div>
```
https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=9875

Specifically if there appears to be one JS library which uses this pattern (an zero height element which would trigger overflow if we added padding universally). I need to study this library some more to determine if making it scrollable will cause any issues for it.

The next step is likely going to be try to ship adding inline-end padding for the *non-empty* case. E.g.
```
<!DOCTYPE html>
<div style="overflow: scroll; display: block; padding: 10px; border: solid; width: 100px; height: 100px;">
  <div style="min-width: 90px; height: 100px; background: lime;"></div>
</div>
```
will be the same as
```
<!DOCTYPE html>
<div style="overflow: scroll; display: flex; padding: 10px; border: solid; width: 100px; height: 100px;">
  <div style="min-width: 90px; height: 100px; background: lime;"></div>
</div>
```

This is likely going to be web compatible as the UseCounter is lower than when we made the similar change for flexbox.

TL;DR at the very least we can add inline-end padding to block boxes which already scroll. But I'd like to keep pushing to try and add this universally.

**Comment by bfgeek** [comment]
Just as a small FYI - we've shipped mostly[*] the expected behaviour in Chromium 99.
[*] Except for the zero inflow-bounds case - this is heading out in 101 - we analyzed the library which was causing the majority of this behaviour, and doesn't affect it.
I'll Adgenda+ this in a couple of months once we are in the final state.

**Comment by bfgeek** [comment]
M101 is rolling at the moment for us. I added the tests for the padding edge behaviour in:

https://wpt.fyi/results/css/css-overflow/scrollable-overflow-padding.html?label=master&label=experimental&aligned&q=css-overflow

All of the tests in that file roughly have this form:

```html
<style>
    scrollable-container {
        display: block;
        width: 100px;
        height: 100px;
        overflow: scroll;
        padding: 10px 5px;
    }

    scrollable-container > div {
        width: 200px; height: 0; /* The child has zero area - so shouldn't directly contribute to scrollable overflow. */
        position: relative; left: -1000px; /* Additionally it is rel-pos'd before the scroll origin. */
        margin-right: 50px; /* end margin to push out the padding edge. */
    }
</style>
<scrollable-container data-expected-scroll-width="260">
    <div></div>
 </scrollable-container>
```

The test does this for various writing-mode + direction combinations + block/inline level children.

Currently all the browsers agree on the block direction scrollable overflow tests. The tests check that the behaviour is the same in the inline direction.

Additionally one this I also wanted to clarify was that there was some discussion in the issue that these two cases should behave the same:
```html
<div style="width: 100px; height: 100px; overflow: scroll; padding: 10px;">
   veryveryverylongword
</div>
<div style="width: 100px; height: 100px; overflow: scroll; padding: 10px;">
   <div>veryveryverylongword</div>
</div>
```

These shouldn't behave the same as the alignment rectangle is different for each. E.g. in the 2nd case the <div> is ~100px, rather than the width of the longest word.

If the `<div>` had something like:
```html
<div style="width: 100px; height: 100px; overflow: scroll; padding: 10px;">
   <div style="min-width: min-content;">veryveryverylongword</div>
</div>
```
Then the test cases would behave the same.

Ian

**Comment by bakura10** [comment]
Hi everyone,

I'm allowing to comment on this issue to have a clarification about this issue.

When working with grid whose content is overflowing, applying a padding is now honored in the scrolling container (which is what as a developer I expect), however this only works if columns are sized in PX. As soon as elements as expressed as a percentage of the parent, this fails and we have here as well an unintuitive behavior.

The issue can be reproduced here: https://codepen.io/bakura10/pen/VwQJEJL

Is this expected as per spec, or is this an implementation issue of browsers (speaking about that, Safari gives a completely different result from other browsers).

**Comment by tabatkins** [comment]
That looks pretty weird! Seems to be an impl bug - are you getting it in multiple browsers?

**Comment by bakura10** [comment]
Chrome and Firefox has a consistent result in not showing the padding end, while Safari has a completely odd output. I will send a ticket to each vendor then :).

**Comment by bfgeek** [comment]
I left a comment at https://bugs.chromium.org/p/chromium/issues/detail?id=1339124#c1 the percentage column case is working as expected - Ian
